llvm/lib/IR/RepoTicket.cpp

```diff
@@ -116,78 +125,135 @@ static GODigestState accumulateGOFinalDigestOrInitialDigestAndGetGODigestState(
   return accumulateGOInitialDigestAndGetGODigestState(GO, GOHash, GOIMap);
 }
 
-// Update the GO's hash value by adding the hash of its dependents.
-template <typename Function>
-static bool updateDigestUseDependenciesAndContributions(
-    const GlobalObject *GO, MD5 &GOHash, GOStateMap &Visited, GOInfoMap &GOIMap,
-    Function AccumulateGODigest) {
+// Update the digest of an invidual GO incorporating the hashes of all its
+// dependencies and contributions.
+// \param GO The global object whose digest is to be computed.
+// \param Visited A map of GO to a unique number in the dependencies or
+//        contributions graph.
+// \param GOIMap A map of GO to GOInfo
+// \param GOHashCache Used to record memoized hashes.
+// \param AccumulateGODigest A function is used to accumulate the GO's digest.
+// \returns a tuple containing the state information which includes 1) return
+// true if the module M has been changed, 2) a numerical identifier which will
+// be used if GO loops back to itself in future and 3) the GO's digest.
+template <typename FunctionType>
+static auto updateDigestUseDependenciesAndContributions(
+    const GlobalObject *GO, GOStateMap &Visited, GOInfoMap &GOIMap,
+    MemoizedHashes &GOHashCache, FunctionType AccumulateGODigest)
+    -> std::tuple<bool, size_t, DigestType> {
   if (GO->isDeclaration())
-    return false;
+    return std::make_tuple(false, std::numeric_limits<size_t>::max(),
+                           NullDigest);
+
+  const auto NumVisited = Visited.size();
+  LLVM_DEBUG(dbgs() << "Computing hash for \"" << GO->getName() << "\" (#"
+                    << NumVisited << ")\n");
+
+  // If the hash has been memoized, we can return the result immediately.
+  const auto TablePos = GOHashCache.find(GO);
+  if (TablePos != GOHashCache.end()) {
+    ++VisitedCachedGOTimes;
+    LLVM_DEBUG(dbgs() << "Returning pre-computed hash for \"" << GO->getName()
+                      << "\"\n");
+    return std::make_tuple(true, NumVisited, TablePos->second);
+  }
+
+  MD5 GOHash = MD5();
 
   bool Inserted;
   typename GOStateMap::const_iterator StateIt;
-  std::tie(StateIt, Inserted) = Visited.try_emplace(GO, Visited.size());
+  std::tie(StateIt, Inserted) = Visited.try_emplace(GO, NumVisited);
   if (!Inserted) {
-    // If GO is visited, use the letter 'R' as the marker and use its state as
-    // the value.
-    GOHash.update('R');
-    GOHash.update(StateIt->second);
-    return true;
+    assert(NumVisited > StateIt->second);
+    // If GO is visited, mark as a Backref and use its state as the value.
+    GOHash.update(static_cast<char>(Tags::Backref));
+    GOHash.update(NumVisited - StateIt->second - 1U);
+    LLVM_DEBUG(dbgs() << "Hashing back reference to #" << StateIt->second
+                      << "\n");
+    MD5::MD5Result Digest;
```

Comment:
Can this by DigestType instead of MD5::MD5Result?

Reply:
Yes, that would be good. We should try to reduce the code’s coupling to MD5: see issue #68.

Reply:
Fixed, thanks.